### PR TITLE
samples: nrf9160: http_modem_update: add sample

### PR DESCRIPTION
**Pull request by lemrey**
I have developed this sample that's basically an integration platform for the `download_client` library and `bsdlib`. I have found it useful to test several features:
- HTTP and TLS including certificate provisioning
- `download_client` library and its compatibility with `bsdlib` versions
- `bsdlib` DFU socket and modem DFU
- `bsdlib` TCP sockets

I think this can be useful for others, and probably for myself in the future, so instead of letting it bitrot in my repo I am opening this PR to push it here. I'd like to get some feedback, mainly as to where to put it, since it is not a proper sample that does something on its own but more of an interactive tool, and what is missing.

FYI: @osaether @sigvartmh

### CODE REVIEW

**Comment by rlubos** [comment]
Soo, how does this relate to #1138? @lemrey @sigvartmh Two alternative approaches to modem DFU?

**Comment by sigvartmh** [comment]
As I understand it this PR exposes the functionality as a CLI through the console. While what I did in #1138 was to integrate this functionality into the `fota_download` (subsys/library). I think both are valid use cases testing/sample. I just don't know what to call this.

File: samples/nrf9160/http_modem_update/src/main.c [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+static nrf_dfu_err_t modem_error_get(void)
+{
+	int rc;
+	nrf_dfu_err_t err;
+	nrf_socklen_t len;
+
+	len = sizeof(err);
+	rc = nrf_getsockopt(fd, NRF_SOL_DFU, NRF_SO_DFU_ERROR, &err, &len);
+	if (rc) {
+		LOG_ERR("Unable to fetch modem error, errno %d", errno);
```
[the remark this answers is not in the document]
Comment (jbrzozoski) [reply samples/nrf9160/http_modem_update/src/main.c]: Sorry for reviving a stale comment, but if your goal is to have the output go to an interactive user shell, I highly recommend shell_print().  Our application waffled around between printk and logger functions while we were using the UART as our only point of interaction, but the moment we added other shell backends it all broke and we had to slog through fixing all of the shell commands to use shell_print.  It's guaranteed to go to the correct interface for the current shell, which none of the others are.  Plus, once you've got shell context passed everywhere, you can more easily make interactive shell commands using `shell->iface->api->read`, which is how we did a xmodem-esque download handler in our shell.

[the remark this answers is not in the document]
Comment (jbrzozoski) [reply samples/nrf9160/http_modem_update/src/main.c]: What I'm trying to hint at is that being able to download firmware via XMODEM from the shell could be useful :)

**Comment by carlescufi** [comment]
Note: This is an automated message.
@lemrey this Pull Request has not been updated for more than a month, please close it if not applicable anymore.

**Pull request closed without merging**